# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsSplittableFrame.cpp
# Commit: 0186165ee1d8
# Full Hash: 0186165ee1d84f56362ba16bbda4eefa1d47fd60
# Author: Mats Palmgren <mats@mozilla.com>
# Date: 2019-11-26 05:19:55
# Regressor Bug: 471015
# File Overlap Count: 1
# Description:
#   Bug 471015 - [css-break] Implement <fieldset> fragmentation.  r=TYLin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D53709
# ==============================================================================

diff -r 9cf1851c76dc -r 0186165ee1d8 layout/generic/nsSplittableFrame.cpp
--- a/layout/generic/nsSplittableFrame.cpp	Mon Nov 25 15:47:38 2019 +0000
+++ b/layout/generic/nsSplittableFrame.cpp	Mon Nov 25 18:05:16 2019 +0000
@@ -11,6 +11,7 @@
 
 #include "nsSplittableFrame.h"
 #include "nsContainerFrame.h"
+#include "nsFieldSetFrame.h"
 #include "nsIFrameInlines.h"
 
 using namespace mozilla;
@@ -205,6 +206,18 @@
 
   bSize -= aConsumedBSize;
 
+  // nsFieldSetFrame's inner frames are special since some of their content-box
+  // BSize may be consumed by positioning it below the legend.  So we always
+  // report zero for true overflow containers here.
+  // XXXmats: hmm, can we fix this so that the sizes actually adds up instead?
+  if (IS_TRUE_OVERFLOW_CONTAINER(this) &&
+      Style()->GetPseudoType() == PseudoStyleType::fieldsetContent) {
+    for (nsFieldSetFrame* fieldset = do_QueryFrame(GetParent()); fieldset;
+         fieldset = static_cast<nsFieldSetFrame*>(fieldset->GetPrevInFlow())) {
+      bSize -= fieldset->LegendSpace();
+    }
+  }
+
   // We may have stretched the frame beyond its computed height. Oh well.
   return std::max(0, bSize);
 }